File: android/src/main/java/io/streamlayer/react/manager/view/StreamLayerViewEventEmitter.kt
package io.streamlayer.react.manager.view

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.WritableMap
import com.facebook.react.common.MapBuilder
import com.facebook.react.uimanager.events.RCTEventEmitter

internal class StreamLayerViewEventEmitter(
  private val reactContext: ReactApplicationContext,
  private val getReactNativeViewId: (() -> Int?)
) {

  fun onRequestStream(id: String) {
    val map = Arguments.createMap()
    map.putString(ARG_ID, id)
    sendEvent(EVENT_REQUEST_STREAM, map)
  }

  fun onLBarStateChanged(slideX: Int, slideY: Int) {
    val map = Arguments.createMap()
    map.putInt(ARG_SLIDE_X, slideX)
    map.putInt(ARG_SLIDE_Y, slideY)
    sendEvent(EVENT_LBAR_STATE_CHANGED, map)
  }

  fun requestAudioDucking(level: Float) {
    val map = Arguments.createMap()
    map.putDouble(ARG_LEVEL, level.toDouble())
    sendEvent(EVENT_REQUEST_AUDIO_DUCKING, map)
  }

  fun disableAudioDucking() {
    sendEvent(EVENT_DISABLE_AUDIO_DUCKING, null)
  }

  private fun sendEvent(type: String, event: WritableMap?) {
    getReactNativeViewId()?.let {
      (reactContext.getJSModule(RCTEventEmitter::class.java) as RCTEventEmitter)
        .receiveEvent(it, type, event)
    }
  }

  companion object {

    // arguments
    private const val ARG_ID = "id"
    private const val ARG_SLIDE_X = "slideX"
    private const val ARG_SLIDE_Y = "slideY"
    private const val ARG_LEVEL = "level"

    // events
    private const val EVENT_REQUEST_STREAM = "onNativeRequestStream"
    private const val EVENT_LBAR_STATE_CHANGED = "onNativeLBarStateChanged"
    private const val EVENT_REQUEST_AUDIO_DUCKING = "onNativeRequestAudioDucking"
    private const val EVENT_DISABLE_AUDIO_DUCKING = "onNativeDisableAudioDucking"

    // setup events
    fun setupEvents(): MutableMap<String, Any> = MapBuilder.builder<String, Any>().apply {
      arrayOf(
        EVENT_REQUEST_STREAM,
        EVENT_LBAR_STATE_CHANGED,
        EVENT_REQUEST_AUDIO_DUCKING,
        EVENT_DISABLE_AUDIO_DUCKING
      ).forEach {
        put(it, MapBuilder.of("registrationName", it))
      }
    }.build()
  }

}

File: ios/StreamLayerViewEventEmitter.swift
//
//  StreamLayerViewEventEmitter.swift
//  react-native-streamlayer
//
//  Created by Kirill Kunst on 10.06.2024.
//

import Foundation
import React

@objc(StreamLayerViewEventEmitter)
class StreamLayerViewEventEmitter: RCTEventEmitter {

    public static var shared: StreamLayerViewEventEmitter?

    enum Args: String {
        case id
        case slideX
        case slideY
        case level
    }

    enum Events: String, CaseIterable {
        case onNativeRequestStream
        case onNativeLBarStateChanged
        case onNativeRequestAudioDucking
        case onNativeDisableAudioDucking
    }

    private var hasObservers: Bool = false

    override init() {
        super.init()
        StreamLayerViewEventEmitter.shared = self
    }

    override class func requiresMainQueueSetup() -> Bool {
        true
    }

    override func startObserving() {
        super.startObserving()
        hasObservers = true
    }

    override func stopObserving() {
        super.stopObserving()
        hasObservers = false
    }

    // MARK: - Events

    func onRequestStream(id: String) {
        guard hasObservers else { return }
        sendEvent(withName: Events.onNativeRequestStream.rawValue, body: [Args.id.rawValue: id])
    }

    func onLBarStateChanged(slideX: Int, slideY: Int) {
        guard hasObservers else { return }
        sendEvent(withName: Events.onNativeLBarStateChanged.rawValue, body: [
            Args.slideX.rawValue: slideX,
            Args.slideY.rawValue: slideY,
        ])
    }

    func requestAudioDucking(level: Float) {
        guard hasObservers else { return }
        sendEvent(withName: Events.onNativeRequestAudioDucking.rawValue, body: [Args.level.rawValue: level])
    }

    func disableAudioDucking() {
        guard hasObservers else { return }
        sendEvent(withName: Events.onNativeDisableAudioDucking.rawValue, body: nil)
    }

    // MARK: RCTEventEmitter overrides

    override func supportedEvents() -> [String]! {
        Events.allCases.map({ $0.rawValue })
    }
}
